File: Practice works (Homework)/LabWork №6/C++/main.cpp
#include <SFML/Graphics.hpp>
#include <SFML/Window/Event.hpp>
#include <iostream>
#include <vector>
#include <queue>
#include <cmath>
#include <algorithm>
#include <random>
#include <optional>

using namespace std;
using namespace sf;

// Размер поля 10x10, так как вариантный лабиринт имеет сетку 10 на 10
const int GRID_SIZE = 10;

// Размер одной клетки в пикселях
const int CELL_SIZE = 60;

// Размер окна программы
const int WINDOW_SIZE = GRID_SIZE * CELL_SIZE;

// Типы клеток на поле
enum CellType {
    EMPTY,      // свободная клетка
    OBSTACLE,   // препятствие
    HEAVY,      // клетка с повышенной стоимостью прохождения
    START,      // стартовая клетка
    END,        // конечная клетка
    PATH,       // найденный путь
    VISITED,    // обработанная клетка, closed list
    OPEN        // клетка в open list
};

// Структура одной клетки поля
struct Cell {
    int x, y;           // координаты клетки
    CellType type;      // тип клетки
    int f, g, h;        // значения для алгоритма A*: f = g + h
    int weight;         // стоимость прохождения клетки
    Cell* parent;       // указатель на предыдущую клетку для восстановления пути

    Cell(int x, int y) : x(x), y(y), type(EMPTY), f(0), g(0), h(0), weight(1), parent(nullptr) {}
};

// Сравнение клеток по значению f.
// Нужно для priority_queue, чтобы выбирать клетку с минимальным f.
struct CompareF {
    bool operator()(Cell* const& c1, Cell* const& c2) {
        return c1->f > c2->f;
    }
};

// Эвристика — манхэттенское расстояние.
// Используется для примерной оценки расстояния от текущей клетки до цели.
int heuristic(int x1, int y1, int x2, int y2) {
    return abs(x1 - x2) + abs(y1 - y2);
}

// Проверка, что координаты находятся внутри поля
bool isValid(int x, int y) {
    return x >= 0 && x < GRID_SIZE && y >= 0 && y < GRID_SIZE;
}

// Функция отрисовки поля
void drawGrid(RenderWindow& window, vector<vector<Cell>>& grid) {
    window.clear(Color::White);

    // Проходим по всем клеткам поля
    for (int i = 0; i < GRID_SIZE; ++i) {
        for (int j = 0; j < GRID_SIZE; ++j) {
            // Создаём квадрат для отображения клетки
            RectangleShape cellShape({ (float)CELL_SIZE - 1.f, (float)CELL_SIZE - 1.f });
            cellShape.setPosition({ (float)(i * CELL_SIZE), (float)(j * CELL_SIZE) });

            // Цвет клетки зависит от её типа
            switch (grid[i][j].type) {
            case EMPTY:
                cellShape.setFillColor(Color::White);
                break;
            case OBSTACLE:
                cellShape.setFillColor(Color::Black);
                break;
            case HEAVY:
                cellShape.setFillColor(Color(210, 180, 140)); // песочный цвет
                break;
            case START:
                cellShape.setFillColor(Color(255, 165, 0)); // оранжевый
                break;
            case END:
                cellShape.setFillColor(Color(64, 224, 208)); // бирюзовый
                break;
            case VISITED:
                cellShape.setFillColor(Color::Red);
                break;
            case OPEN:
                cellShape.setFillColor(Color::Green);
                break;
            case PATH:
                cellShape.setFillColor(Color(128, 0, 128)); // фиолетовый
                break;
            }

            window.draw(cellShape);
        }
    }

    window.display();
}

// Загрузка фиксированного лабиринта для варианта 8
void loadVariant8(vector<vector<Cell>>& grid, Cell*& start, Cell*& end) {
    // Сначала очищаем поле и создаём новые пустые клетки
    for (int i = 0; i < GRID_SIZE; ++i) {
        for (int j = 0; j < GRID_SIZE; ++j) {
            grid[i][j] = Cell(i, j);
        }
    }

    // Устанавливаем стартовую точку
    start = &grid[0][3];
    start->type = START;

    // Устанавливаем конечную точку
    end = &grid[9][9];
    end->type = END;

    // Координаты препятствий для варианта 8
    vector<pair<int, int>> obstacles = {
        {0,0}, {1,0},{2,0}, {3,0}, {4,0}, {6,0},
        {0,1}, {1,1}, {5,1},
        {1,2},
        {2,3}, {4,3}, {5,3}, {7,3}, {8,3},
        {1,4}, {3,4}, {5,4}, {6,4}, {7,4},
        {6,5}, {7,5}, {8,5}, {9,5},
        {0,6}, {4,6}, {5,6},
        {0,7}, {5,7}, {6,7}, {7,7}, {9,7},
        {8,8},
        {2,9}
    };

    // Расставляем препятствия на поле
    for (auto obs : obstacles) {
        grid[obs.first][obs.second].type = OBSTACLE;
    }
}

// Основная функция алгоритма A*
void aStar(vector<vector<Cell>>& grid, Cell* start, Cell* end, RenderWindow& window) {
    // Открытый список: клетки, которые нужно рассмотреть
    priority_queue<Cell*, vector<Cell*>, CompareF> open_list;

    // Закрытый список: клетки, которые уже были обработаны
    vector<vector<bool>> closed_list(GRID_SIZE, vector<bool>(GRID_SIZE, false));

    // Вспомогательный массив: показывает, находится ли клетка в open_list
    vector<vector<bool>> in_open_list(GRID_SIZE, vector<bool>(GRID_SIZE, false));

    // Начальная клетка имеет стоимость g = 0
    start->g = 0;

    // h — эвристическая оценка расстояния от старта до цели
    start->h = heuristic(start->x, start->y, end->x, end->y);

    // f — итоговая оценка клетки
    start->f = start->g + start->h;

    // Добавляем стартовую клетку в open list
    open_list.push(start);
    in_open_list[start->x][start->y] = true;

    // Возможные направления движения: вверх, вниз, влево, вправо
    int dx[] = { 0, 0, -1, 1 };
    int dy[] = { -1, 1, 0, 0 };

    // Пока есть клетки, которые можно рассмотреть
    while (!open_list.empty()) {
        // Берём клетку с минимальным значением f
        Cell* current = open_list.top();
        open_list.pop();

        // После извлечения клетка больше не считается находящейся в open list
        in_open_list[current->x][current->y] = false;

        // Если достигли конечной клетки, восстанавливаем путь
        if (current->x == end->x && current->y == end->y) {
            Cell* temp = current->parent;

            // Идём от конца к старту по указателям parent
            while (temp != nullptr && temp != start) {
                temp->type = PATH;
                temp = temp->parent;
            }

            return;
        }

        // Переносим текущую клетку в closed list
        closed_list[current->x][current->y] = true;

        // Отмечаем обработанную клетку красным цветом
        if (current != start && current != end) {
            current->type = VISITED;
        }

        // Проверяем четырёх соседей текущей клетки
        for (int i = 0; i < 4; ++i) {
            int nx = current->x + dx[i];
            int ny = current->y + dy[i];

            // Проверяем, что сосед находится внутри поля
            if (isValid(nx, ny)) {
                Cell* neighbor = &grid[nx][ny];

                // Препятствия и уже обработанные клетки пропускаем
                if (neighbor->type == OBSTACLE || closed_list[nx][ny]) {
                    continue;
                }

                // Новая стоимость пути до соседней клетки.
                // Здесь учитывается вес клетки.
                int tentative_g = current->g + neighbor->weight;

                // Если сосед ещё не в open list или найден более короткий путь
                if (!in_open_list[nx][ny] || tentative_g < neighbor->g) {
                    neighbor->g = tentative_g;
                    neighbor->h = heuristic(nx, ny, end->x, end->y);
                    neighbor->f = neighbor->g + neighbor->h;
                    neighbor->parent = current;

                    // Если соседней клетки ещё нет в open list, добавляем её
                    if (!in_open_list[nx][ny]) {
                        open_list.push(neighbor);
                        in_open_list[nx][ny] = true;

                        // Отмечаем клетку зелёным, если это не конечная точка
                        if (neighbor != end) {
                            neighbor->type = OPEN;
                        }
                    }
                }
            }
        }

        // Перерисовываем поле после каждого шага алгоритма
        drawGrid(window, grid);
    }
}

int main() {
    // Создаём окно SFML
    RenderWindow window(VideoMode({ (unsigned int)WINDOW_SIZE, (unsigned int)WINDOW_SIZE }), "A* Algorithm - Variant 8");

    // Создаём двумерную сетку клеток
    vector<vector<Cell>> grid(GRID_SIZE, vector<Cell>(GRID_SIZE, Cell(0, 0)));

    Cell* start = nullptr;
    Cell* end = nullptr;

    // Загружаем лабиринт варианта 8
    loadVariant8(grid, start, end);

    // Основной цикл программы
    while (window.isOpen()) {
        while (const std::optional<Event> event = window.pollEvent()) {
            // Закрытие окна
            if (event->is<Event::Closed>()) {
                window.close();
            }
            else if (const auto* keyPress = event->getIf<Event::KeyPressed>()) {
                // SPACE — запуск алгоритма A*
                if (keyPress->code == Keyboard::Key::Space) {
                    aStar(grid, start, end, window);
                }

                // V — возврат к варианту 8
                if (keyPress->code == Keyboard::Key::V) {
                    loadVariant8(grid, start, end);
                }

                // R — генерация случайного поля с препятствиями
                if (keyPress->code == Keyboard::Key::R) {
                    // Очищаем поле
                    for (int i = 0; i < GRID_SIZE; i++) {
                        for (int j = 0; j < GRID_SIZE; j++) {
                            grid[i][j] = Cell(i, j);
                        }
                    }

                    // Случайно выбираем старт и конец
                    start = &grid[rand() % GRID_SIZE][rand() % GRID_SIZE];
                    end = &grid[rand() % GRID_SIZE][rand() % GRID_SIZE];

                    // Проверяем, чтобы старт и конец не совпадали
                    while (start == end) {
                        end = &grid[rand() % GRID_SIZE][rand() % GRID_SIZE];
                    }

                    start->type = START;
                    end->type = END;

                    // Добавляем случайные препятствия
                    for (int i = 0; i < 20; i++) {
                        int rx = rand() % GRID_SIZE;
                        int ry = rand() % GRID_SIZE;

                        if (grid[rx][ry].type == EMPTY) {
                            grid[rx][ry].type = OBSTACLE;
                        }
                    }
                }

                // W — генерация случайной карты с весами
                if (keyPress->code == Keyboard::Key::W) {
                    // Очищаем поле
                    for (int i = 0; i < GRID_SIZE; i++) {
                        for (int j = 0; j < GRID_SIZE; j++) {
                            grid[i][j] = Cell(i, j);
                        }
                    }

                    // Случайно выбираем стартовую и конечную клетки
                    start = &grid[rand() % GRID_SIZE][rand() % GRID_SIZE];
                    end = &grid[rand() % GRID_SIZE][rand() % GRID_SIZE];

                    // Старт и конец не должны совпадать
                    while (start == end) {
                        end = &grid[rand() % GRID_SIZE][rand() % GRID_SIZE];
                    }

                    start->type = START;
                    end->type = END;

                    // Добавляем клетки с повышенной стоимостью прохождения
                    for (int i = 0; i < 30; i++) {
                        int rx = rand() % GRID_SIZE;
                        int ry = rand() % GRID_SIZE;

                        if (grid[rx][ry].type == EMPTY) {
                            grid[rx][ry].type = HEAVY;
                            grid[rx][ry].weight = 5;
                        }
                    }
                }
            }
        }

        // Постоянно отображаем текущее состояние поля
        drawGrid(window, grid);
    }

    return 0;
}

File: Practice works (Homework)/LabWork №6/Python/main.py
import pygame
import random
from queue import PriorityQueue

pygame.init()

WIDTH = 600
GRID_SIZE = 10
CELL_SIZE = WIDTH // GRID_SIZE
WIN = pygame.display.set_mode((WIDTH, WIDTH))
pygame.display.set_caption("Алгоритм A* - Вариант 8")

# Цвета
WHITE = (255, 255, 255)
BLACK = (0, 0, 0)  # Препятствие
ORANGE = (255, 165, 0)  # Старт
TURQUOISE = (64, 224, 208)  # Конец
GREEN = (0, 255, 0)  # В очереди (Open list)
RED = (255, 0, 0)  # Посещена (Closed list)
PURPLE = (128, 0, 128)  # Итоговый путь
GREY = (128, 128, 128)  # Линии сетки
BROWN = (210, 180, 140)  # Клетка с весом (труднопроходимая)


class Cell:
    def __init__(self, x, y):
        self.x_idx = x
        self.y_idx = y
        self.x_coord = x * CELL_SIZE
        self.y_coord = y * CELL_SIZE
        self.color = WHITE
        self.weight = 1
        self.neighbors = []

    def get_pos(self):
        return self.x_idx, self.y_idx

    def is_obstacle(self):
        return self.color == BLACK

    def reset(self):
        self.color = WHITE
        self.weight = 1

    def make_start(self):
        self.color = ORANGE
        self.weight = 1

    def make_end(self):
        self.color = TURQUOISE
        self.weight = 1

    def make_obstacle(self):
        self.color = BLACK
        self.weight = 1

    def make_heavy(self):
        self.color = BROWN
        self.weight = 5

    def make_open(self):
        self.color = GREEN

    def make_closed(self):
        self.color = RED

    def make_path(self):
        self.color = PURPLE

    def draw(self, win):
        pygame.draw.rect(win, self.color, (self.x_coord, self.y_coord, CELL_SIZE, CELL_SIZE))

    def update_neighbors(self, grid):
        self.neighbors = []
        # Вниз (y + 1)
        if self.y_idx < GRID_SIZE - 1 and not grid[self.x_idx][self.y_idx + 1].is_obstacle():
            self.neighbors.append(grid[self.x_idx][self.y_idx + 1])
        # Вверх (y - 1)
        if self.y_idx > 0 and not grid[self.x_idx][self.y_idx - 1].is_obstacle():
            self.neighbors.append(grid[self.x_idx][self.y_idx - 1])
        # Вправо (x + 1)
        if self.x_idx < GRID_SIZE - 1 and not grid[self.x_idx + 1][self.y_idx].is_obstacle():
            self.neighbors.append(grid[self.x_idx + 1][self.y_idx])
        # Влево (x - 1)
        if self.x_idx > 0 and not grid[self.x_idx - 1][self.y_idx].is_obstacle():
            self.neighbors.append(grid[self.x_idx - 1][self.y_idx])


def h(p1, p2):
    x1, y1 = p1
    x2, y2 = p2
    return abs(x1 - x2) + abs(y1 - y2)


def reconstruct_path(came_from, current, draw):
    while current in came_from:
        current = came_from[current]
        current.make_path()
        draw()


def a_star_algorithm(draw, grid, start, end):
    count = 0
    open_set = PriorityQueue()
    open_set.put((0, count, start))
    came_from = {}

    g_score = {cell: float("inf") for row in grid for cell in row}
    g_score[start] = 0

    f_score = {cell: float("inf") for row in grid for cell in row}
    f_score[start] = h(start.get_pos(), end.get_pos())

    open_set_hash = {start}

    while not open_set.empty():

        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()

        current = open_set.get()[2]
        open_set_hash.remove(current)

        if current == end:
            reconstruct_path(came_from, end, draw)
            end.make_end()
            start.make_start()
            return True

        for neighbor in current.neighbors:
            temp_g_score = g_score[current] + neighbor.weight

            if temp_g_score < g_score[neighbor]:
                came_from[neighbor] = current
                g_score[neighbor] = temp_g_score
                f_score[neighbor] = temp_g_score + h(neighbor.get_pos(), end.get_pos())

                if neighbor not in open_set_hash:
                    count += 1
                    open_set.put((f_score[neighbor], count, neighbor))
                    open_set_hash.add(neighbor)
                    neighbor.make_open()

        draw()
        if current != start:
            current.make_closed()

    return False


def make_grid():
    grid = []
    for i in range(GRID_SIZE):
        grid.append([])
        for j in range(GRID_SIZE):
            cell = Cell(i, j)  # i = x, j = y
            grid[i].append(cell)
    return grid


def draw_grid_lines(win):
    for i in range(GRID_SIZE):
        pygame.draw.line(win, GREY, (0, i * CELL_SIZE), (WIDTH, i * CELL_SIZE))
        pygame.draw.line(win, GREY, (i * CELL_SIZE, 0), (i * CELL_SIZE, WIDTH))


def draw(win, grid):
    for row in grid:
        for cell in row:
            cell.draw(win)
    draw_grid_lines(win)
    pygame.display.update()


def load_variant_8(grid):
    # Точные координаты Варианта 8
    obstacles = [
        (0, 0), (1, 0), (2, 0), (3, 0), (4, 0), (6, 0),
        (0, 1), (1, 1), (5, 1),
        (1, 2), (6, 2), (7, 2),
        (2, 3), (4, 3), (5, 3), (7, 3), (8, 3),
        (1, 4), (3, 4), (5, 4), (6, 4), (7, 4),
        (6, 5), (7, 5), (8, 5), (9, 5),
        (0, 6), (4, 6), (5, 6),
        (0, 7), (5, 7), (6, 7), (7, 7), (9, 7),
        (8, 8),
        (2, 9)
    ]
    for x, y in obstacles:
        grid[x][y].make_obstacle()


def main():
    grid = make_grid()

    # Старт X=0, Y=3 | Финиш X=9, Y=9
    start = grid[0][3]
    end = grid[9][9]

    load_variant_8(grid)
    start.make_start()
    end.make_end()

    run = True
    while run:
        draw(WIN, grid)
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                run = False

            if event.type == pygame.KEYDOWN:
                if event.key == pygame.K_SPACE:
                    for row in grid:
                        for cell in row:
                            cell.update_neighbors(grid)
                    a_star_algorithm(lambda: draw(WIN, grid), grid, start, end)

                if event.key == pygame.K_c:
                    for row in grid:
                        for cell in row:
                            if cell.color in [GREEN, RED, PURPLE]:
                                cell.reset()
                    start.make_start()
                    end.make_end()

                if event.key == pygame.K_1:
                    start = grid[0][3]
                    end = grid[9][9]
                    for row in grid:
                        for cell in row:
                            cell.reset()
                    load_variant_8(grid)
                    start.make_start()
                    end.make_end()

                if event.key == pygame.K_2:
                    for row in grid:
                        for cell in row:
                            cell.reset()
                    start = grid[random.randint(0, 9)][random.randint(0, 9)]
                    end = grid[random.randint(0, 9)][random.randint(0, 9)]
                    while start == end:
                        end = grid[random.randint(0, 9)][random.randint(0, 9)]
                    start.make_start()
                    end.make_end()
                    for _ in range(20):
                        x, y = random.randint(0, 9), random.randint(0, 9)
                        cell = grid[x][y]
                        if cell != start and cell != end:
                            cell.make_obstacle()

                if event.key == pygame.K_3:
                    for row in grid:
                        for cell in row:
                            cell.reset()
                    start = grid[random.randint(0, 9)][random.randint(0, 9)]
                    end = grid[random.randint(0, 9)][random.randint(0, 9)]
                    while start == end:
                        end = grid[random.randint(0, 9)][random.randint(0, 9)]
                    start.make_start()
                    end.make_end()
                    for _ in range(30):
                        x, y = random.randint(0, 9), random.randint(0, 9)
                        cell = grid[x][y]
                        if cell != start and cell != end:
                            cell.make_heavy()

    pygame.quit()


if __name__ == "__main__":
    main()